Zephyr: Add automatic filesystem creation.

bikeNomad opened the pull request:
### Summary

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->
Existing C code in `main.c` only mounts a filesystem if one exists, and
doesn't do anything if the `storage` partition is not formatted.

This commit moves the mounting logic from `main.c` to `modules/_boot.py` and
adds the formatting of a previously unformatted partition if the mount
fails.

To enable this, you must have `CONFIG_MICROPY_FROZEN_MODULES` enabled in your `.conf` file,
and have `CONFIG_MICROPY_FROZEN_MANIFEST` set to point to a `manifest.py` that includes
`modules` (which `boards/manifest.py` already does).
For example, here's what I added for the Raspberry Pi Pico to test:

```
# boards/rpi_pico.conf additions
CONFIG_MICROPY_FROZEN_MODULES=y
CONFIG_MICROPY_FROZEN_MANIFEST="boards/manifest.py"
```

### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->
This was tested on a XIAO NRF52480 SENSE board (with the code from #17679), and also with a Raspberry Pi Pico,
after modifying its configuration as described above.

### Trade-offs and Alternatives

<!-- If the Pull Request has some negative impact (i.e. increased code size)
     then please explain why you think the trade-off improvement is worth it.
     If you can think of alternative ways to do this, please explain that here too.

     Delete this heading if not relevant (i.e. small fixes) -->
There was logic in `main.c` that would mount an SD card on `/sd` if
`CONFIG_DISK_DRIVER_SDMMC` was defined, but none of the existing boards
define this (`CONFIG_DISK_ACCESS` is only defined for the `mimxrt1050_evk`,
but this board doesn't have `CONFIG_DISK_DRIVER_SDMMC` defined). Because
of this, the current PR doesn't include the SDMMC logic from `main.c`.

This code could be added back to `main.c` if desired, but I didn't have any
supported board to test with.

bikeNomad committed:
zephyr: Mount all disks and flash partition, formatting if necessary.

Existing C code in `main.c` only mounts a flash filesystem if one exists,
and doesn't do anything if the 'storage' partition is not formatted.

This commit moves the mounting logic from `main.c` to frozen code using
`modules/_boot.py` and adds the formatting of a previously unformatted
partition if the mount fails.

Every available disk (in the newly added `DiskAccess.disks` tuple) will be
mounted on separate mount points (if they're formatted), and the 'storage'
flash partition (if any) will be mounted on /flash (and will be formatted
as LFS2 if necessary).

Also, `sys.path` will be updated with appropriate 'lib' subdirectories for
each mounted filesystem.

The current working directory will be changed to the last `DiskAccess.disk`
mounted, or to /flash if no disks were mounted.

Then `boot.py` and `main.py` will be executed from the current working
directory if they exist.

Thanks to @VynDragon for the logic in `zephyr/zephyr_storage.c`.

Signed-off-by: Ned Konz <ned@metamagix.tech>

VynDragon commented:
You will also need to re-add a way to automount /sd i think

bikeNomad commented:
I found a dev board with an SD controller and slot. I've got the mounting of both the SD and flash working; the SD is being mounted with the old code and the flash is being mounted with the code in `_boot.py`.

The question is this: If we have both `/sd` and `/flash` mounted, which one should be the initial working directory, and which one's `/lib` directory should come first in the `sys.path`?

This will also affect which `boot.py` and `main.py` get executed.

In my current code, if both exist the SD will have precedence.

VynDragon commented:
I'll look into getting you a code snippet to have the SD available in the python side like Flash Area so it can be coherent

VynDragon commented:
[sd.patch](https://github.com/user-attachments/files/22699926/sd.patch)
Here you go, this should do the same for the disk access entries than for the flash areas.
Makes think storing the disk names and ID directly in the class is not such a bright idea.

VynDragon commented:
[sd.patch](https://github.com/user-attachments/files/22702596/sd.patch)
Here is a much better one that fixes the issue of having all the names directly in the class, here it's nice const tables.

bikeNomad commented:
> [sd.patch](https://github.com/user-attachments/files/22702596/sd.patch) Here is a much better one that fixes the issue of having all the names directly in the class, here it's nice const tables.

@VynDragon Thanks! I've changed my code to use your new method.

VynDragon commented on ports/zephyr/modules/_boot.py:
```diff
@@ -2,13 +2,13 @@
 import sys
 import os
-from zephyr import FlashArea
+from zephyr import FlashArea, DiskAccess
```
You need to check if the modules exist before doing the relevant things with them, as they will be missing if not provided

VynDragon commented:
Version with FS support (only sharing, for future)
[_boot.py](https://github.com/user-attachments/files/22711924/_boot.py)

dpgeorge commented on ports/zephyr/main.c:
```diff
@@ -152,10 +120,16 @@ int real_main(void) {
+
 
-    #if MICROPY_MODULE_FROZEN || MICROPY_VFS
+    #if MICROPY_MODULE_FROZEN_MPY || MICROPY_VFS
```
I think this can stay as `MICROPY_MODULE_FROZEN`.

dpgeorge commented on ports/zephyr/main.c:
```diff
@@ -152,10 +120,16 @@ int real_main(void) {
+    // Mount and/or create the filesystem
+    pyexec_frozen_module("_boot.py", false);
+    #else
     vfs_init();
```
This function has been removed, so this line must go.

dpgeorge committed:
tools/ci.sh: Init micropython-lib submodule for zephyr build.

The zephyr port doesn't have a Makefile so can't run `make submodules`.
Instead they must be explicitly initialized.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented on ports/zephyr/modules/_boot.py:
```diff
@@ -0,0 +1,73 @@
+
+_FLASH = const("/flash")
+_FLASH_LIB = const("/flash/lib")
+_STORAGE_KEY = const("Storage")
```
Why did you change this from `storage`?  It no longer works because almost all zephyr boards use lowercase "storage" for the label name of the storage partition.

bikeNomad replied:
Sorry about that; I tested with the XIAO NRF52840 and its DTS had "Storage". I'll fix it there.

bikeNomad committed:
zephyr: Add erase block size to FlashArea.areas entries.

This commit changes the values in the `FlashArea.areas` dictionary from
simple integer IDs to (ID, erase-block-size) tuples.

`_boot.py` was changed to use the newly available erase block size.

Signed-off-by: Ned Konz <ned@metamagix.tech>

dpgeorge reviewed (approved):
This looks really good now, having the block size as part of the `FlashArea.areas` works well.

Tested on `rpi_pico`, it works for me.
